plugins/net_plugin/net_plugin.cpp

```diff
@@ -333,9 +336,36 @@ namespace eosio {
    constexpr uint32_t packed_transaction_which = fc::get_index<net_message, packed_transaction>(); // see protocol net_message
 
    class connections_manager {
+   public:
+      struct connection_details {
+         std::string host;
+         connection_ptr c;
+         tcp::endpoint active_ip;
+         std::vector<tcp::endpoint> ips;
+         operator const connection_ptr&() const { return c; }
+      };
+
+      using connection_details_index = multi_index_container<
+         connection_details,
+         indexed_by<
+            ordered_unique<
+               tag<struct by_host>,
+               key<&connection_details::host>
+            >,
+            ordered_unique<
+               tag<struct by_connection>,
+               key<&connection_details::c>
+            >,
+            ordered_non_unique<
+               tag<struct by_active_ip>,
+               key<&connection_details::active_ip>
+            >
+         >
+      >;
+   private:
       alignas(hardware_destructive_interference_size)
       mutable std::shared_mutex        connections_mtx;
-      chain::flat_set<connection_ptr>  connections;
+      connection_details_index         connections GUARDED_BY(connections_mtx);
```

Comment:
The GUARDED_BY does not work with `std::shared_lock`. Remove this.

Reply:
Removed.